### ISSUE TITLE
[F#] Implement new Concept Exercise: type-inference

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/fsharp#858]
This issue describes how to implement the `type-inference` concept exercise for the F# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the basics of the Concept of Type Inference in F#.

## Learning objectives

- Know what type-inference is
- Know how type-inference works.
- Know on what values type inference works (including records, tuples and discriminated unions).
- Know about the limitations of type inference and how to work around them.
- Know what automatic generalization is.
- Know what generic functions are.

## Out of scope

- Performance aspects.

## Concepts

The Concepts this exercise unlocks are:

- `type-inference`: know what type-inference is; know how type-inference works; know on what values type inference works; know about the limitations of type inference and how to work around them; know what automatic generalization is.
- `generic-functions`: know what generic functions are.

## Prerequisites

This exercise's prerequisites Concepts are:

- `records`: know what records are.
- `tuples`: know what tuples are.
- `discriminated-unions`: know what discriminated unions are.
- `higher-order-functions`: know how to read function type signatures.

Any data types used in this exercise (e.g. `strings` or `integers`) should also be added as prerequisites.

### Hints

- [Type inference][type-inference]: introduction to type inference.
- [Type inference for functions][type-inference-for-functions]: how type inference works for functions.
- [Automatic generalization][automatic-generalization]: what automatica generalization is and how to use it.

### After

- [Type inference][type-inference]: introduction to type inference.
- [Type inference for functions][type-inference-for-functions]: how type inference works for functions.
- [Automatic generalization][automatic-generalization]: what automatica generalization is and how to use it.
- [Type inference in-depth][type-inference-in-depth]: in-depth look at type-inference.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/fsharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/reference/concepts/type_inference.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/fsharp-analyzer
[representer]: https://github.com/exercism/fsharp-representer
[meta-design]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/design.md
[meta-config.json]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/config.json
[docs-concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[type-inference]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/type-inference
[type-inference-for-functions]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/type-inference#inference-of-parameter-and-return-types
[automatic-generalization]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/type-inference#automatic-generalization
[type-inference-in-depth]: https://fsharpforfunandprofit.com/posts/type-inference/
